### PR TITLE
fix: split subscript ranges and equations on the `|` delimiter

### LINKED ISSUE #7: Successive 2D constant arrays are merged incorrectly
When a model contains two successive 2D constant arrays (after sorting into variable name order by the preprocessor) the equations are merged by the parser. For instance, this model:

```
DimA: A1, A2, A3 ~~|
DimB: B1, B2 ~~|

p[DimA, DimB]=
0, 0;
0, 0;
0, 0;
~~|

q[DimA, DimB]=
1, 1;
1, 1;
1, 1;
~~|
```

shows the following formula in the variable listing for the first equation.

```
= 0,0;0,0;0,0;q[DimA,DimB]=1,1;1,1;1,1;
```

### PR DESCRIPTION
**Pull request by ToddFincannon**
Fixes #7 

See issue #7 for details.

### CODE REVIEW

File: grammar/Model.g4 [hunk 1]
```diff
@@ -5,21 +5,21 @@ import Expr;
...
 // The RHS is a formula expression, a constant list, or a Vensim lookup.
 // The RHS is empty for data equations.
-equation : lhs ( ( ':=' | '==' | '=' ) ( expr | constList ) | lookup )? ;
+unitsDoc : '~' .*? '|' ;
+equation : lhs ( ( ':=' | '==' | '=' ) ( expr | constList ) | lookup )? unitsDoc* ;
 lhs : Id ( '[' subscriptList ']' )? ':INTERPOLATE:'? ( ':EXCEPT:' '[' subscriptList ']' ( ',' '[' subscriptList ']' )* )? ;
 
 // The lexer strips some tokens we are not interested in.
 // The character encoding is given at the start of a Vensim file.
 // The units and documentation sections and group markings are skipped for now.
```
Comment (chrispcampbell) [inline grammar/Model.g4:21]: This comment ("The units and documentation sections and group markings are skipped for now.") needs to be updated now that we're no longer skipping units/docs.

**Review by chrispcampbell — CHANGES_REQUESTED** [review]
@ToddFincannon Hrm, this one seemed reasonable at first glance, but I tried using this fixed version (both with and without your SDE changes for PR 110) to build En-ROADS and it seems to cause a significant performance regression.  Running `sde generate --genc --spec $SPEC enroads`:
```
before this fix:  4.6 seconds
after this fix:  21.4 seconds
```

Any idea what specifically causes the slowdown, or if there's an alternate approach?

File: grammar/Model.g4 (hunk 1 above)
Comment (ToddFincannon) [reply grammar/Model.g4:21]: The `unitsDoc` expression may match a large amount of text, as some of the comments are lengthy. Now that they're not simply skipped, the parser may be taking a lot of time and memory handling them.

I don't see how to resolve the SDEverywhere #109 bug without splitting equations on this expression. Maybe there's a way to include it in the equation and subscriptRange expressions and still discard the contents. I will find out.

**Comment by ToddFincannon** [comment]
From the issue comments:

This parser change resulted in a large increase in code generation time. The cause is probably that unitsDoc is now a parser rule that injects large amounts of text into the parse tree instead of skipping it in the lexer.

We can handle this issue a different way by reverting to the UnitsDoc lexer rule, and simply terminating equations and subscript ranges on the | character. I did some tests and confirmed that the vertical bar character cannot be part of an equation or doc string. The UnitsDoc rule matches everything up to but not including the vertical bar.

This approach requires SDEverywhere to add the ~~| terminator to the end of equations that it generates for delay expansions, etc. We should probably have done this all along. See SDEverywhere issue #119 for a commit with the necessary changes.

File: grammar/Model.g4 (hunk 1 above)
Comment (chrispcampbell) [reply grammar/Model.g4:21]: Now that we're back to skipping `unitsDoc`, the code comment is once again accurate, so my review comment above is now obsolete.  Marking this conversation as resolved.

**Review by chrispcampbell — APPROVED** [review]
@ToddFincannon I tested this fix (via the required SDE branch) with the En-ROADS build and can confirm that code gen performance is back to normal.

Running `sde generate --genc --spec $SPEC enroads`:
```
before this fix:  4.7 seconds (+/- ~0.2 s)
after this fix:  4.7 seconds (+/- ~0.2 s)
```

I will merge this shortly.

**Pull request merged**